accord-core/src/main/java/accord/local/cfk/Updating.java

```diff
@@ -855,22 +870,26 @@ else if (register)
 
                 CommandsForKey result;
                 if (newById == byId) result = new CommandsForKey(cfk, newLoadingPruned, newUnmanaged);
-                else result = new CommandsForKey(cfk.key(), cfk.redundantBefore(), cfk.prunedBefore, newLoadingPruned, newById, newCommittedByExecuteAt, newMinUndecidedById, cfk.maxAppliedWriteByExecuteAt, newUnmanaged);
+                else
+                {
+                    int prunedBeforeById = cfk.prunedBeforeById;
+                    Invariants.checkState(prunedBeforeById < 0 || newById[prunedBeforeById].equals(cfk.prunedBefore()));
+                    result = new CommandsForKey(cfk.key(), cfk.redundantBefore, cfk.bootstrappedAt, newById, newCommittedByExecuteAt, newMinUndecidedById, cfk.maxAppliedWriteByExecuteAt, newLoadingPruned, prunedBeforeById, cfk.safelyPrunedBefore, newUnmanaged);
+                }
 
                 if (loadPruned == NO_TXNIDS)
                     return result;
 
-                return new CommandsForKeyUpdate.CommandsForKeyUpdateWithNotifier(result, new PostProcess.LoadPruned(null, loadPruned));
+                return new CommandsForKeyUpdate.CommandsForKeyUpdateWithPostProcess(result, new PostProcess.LoadPruned(null, loadPruned));
             }
         }
 
-        notifySink.notWaiting(safeStore, safeCommand, cfk.key());
-        return cfk;
+        return new CommandsForKeyUpdate.CommandsForKeyUpdateWithPostProcess(cfk, new PostProcess.NotifyNotWaiting(null, new TxnId[] { safeCommand.txnId() }));
     }
 
     static CommandsForKeyUpdate registerHistorical(CommandsForKey cfk, TxnId txnId)
     {
-        if (txnId.compareTo(cfk.shardRedundantBefore()) < 0)
+        if (txnId.compareTo(cfk.redundantBefore()) < 0)
```

Comment:
yes, from a `CommandsForKey` perspective its contents are redundant if they are shard-redundant for the key.

Reply:
My next patch actually undoes some of this refactoring, bringing back the `RedundantBefore.Entry` because I need to also consider the start and end epochs. So, perhaps I could restore it here as well if you prefer.